### PR TITLE
UnusedPrivateMember: fix false negative when file/class has same named functions

### LINKED ISSUE #1981: UnusedPrivateMember doesn't report the correct warning count
## Expected Behavior
```kotlin
class Test {
    private fun unusedFunction(): Int {
        return 5
    }
}

class Test2 {
    private fun unusedFunction(): Int {
        return 5
    }
}
```

In that code I expect to have two warnings because there are two unused functions

## Observed Behavior
Only one warning is raised. The reason is becase the functions are called the same.

## Steps to Reproduce
You can check this comparison: https://github.com/arturbosch/detekt/compare/master...BraisGabin:unused-private-function

## Context
I found this issue fixing a similar issue with parameters that are called the same in different functions. That issue was easier to fix that this one.

## Your Environment
* Version of detekt used: 1.0.1
* Operating System and version: Mac

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #1981

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt [hunk 1]
```diff
@@ -118,19 +130,12 @@ private class UnusedFunctionVisitor(allowedNames: Regex) : UnusedMemberVisitor(a
...
-
-        if (calledMethodName != null) {
-            callExpressions.add(calledMethodName)
-        }
+        val name = when (expression) {
+            is KtOperationReferenceExpression -> expression.getReferencedName()
+            is KtNameReferenceExpression -> expression.getReferencedName()
+            else -> null
+        } ?: return
+        functionReferences.getOrPut(name) { mutableListOf() }.add(expression)
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: Could these statements be simplified to have a lower complexity? I'm thinking about something like this.

```suggestion
        if (expression is KtOperationReferenceExpression || expression is KtNameReferenceExpression) {
            val name = expression.getReferencedName()
            functionReferences.getOrPut(name) { mutableListOf() }.add(expression)
        }
```

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt [hunk 2]
```diff
@@ -73,19 +73,31 @@ private abstract class UnusedMemberVisitor(protected val allowedNames: Regex) :
...
+    private val functionDeclarations = mutableMapOf<String, MutableList<KtFunction>>()
+    private val functionReferences = mutableMapOf<String, MutableList<KtReferenceExpression>>()
 
     override fun getUnusedReports(issue: Issue): List<CodeSmell> {
-        for (call in callExpressions) {
-            if (functions.isEmpty()) {
-                break
+        return functionDeclarations.flatMap { (name, functions) ->
+            val references = functionReferences[name].orEmpty()
+            val unusedFunctions = if (functions.size > 1 && bindingContext != BindingContext.EMPTY) {
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: `functions.size > 1`

I'm sorry, from the reading the code I don't quite understand why you check this condition.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt [hunk 3]
```diff
@@ -73,19 +73,31 @@ private abstract class UnusedMemberVisitor(protected val allowedNames: Regex) :
...
+                val referenceDescriptors = references.mapNotNull {
+                    it.getResolvedCall(bindingContext)?.resultingDescriptor
+                }
+                functions.filter {
+                    bindingContext[BindingContext.DECLARATION_TO_DESCRIPTOR, it] !in referenceDescriptors
+                }
+            } else {
+                if (references.isEmpty()) functions else emptyList()
+            }
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]:
```suggestion
            } else if (references.isEmpty()) {
                functions
            } else {
                emptyList()
            }
```

Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: [minor] formatting

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt (hunk 1 above)
Comment (t-kameyama) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: No. To call 'getReferencedName()', smart cast is required.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt (hunk 2 above)
Comment (t-kameyama) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: If there is no function with the same name, just checking the name is enough.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt (hunk 1 above)
Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: Thanks for the explanation! :-)
resolved...

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt (hunk 2 above)
Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: Oh yes. Good one!

**Pull request merged**